File: pxpoint/PxPointLogger.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using PxPoint.Correlation;

namespace PxPoint.Logging
{
    /// <summary>
    /// PxPoint-specific logger that integrates with the correlation system
    /// and writes to /tmp/pxpoint-logs/
    /// </summary>
    public class PxPointLogger
    {
        private static readonly Lazy<PxPointLogger> _instance = 
            new Lazy<PxPointLogger>(() => new PxPointLogger());
        
        public static PxPointLogger Instance => _instance.Value;
        
        private bool _initialized = false;
        private string _processType = string.Empty;
        private StreamWriter _fileWriter;
        
        private PxPointLogger() { }
        
        // Initialize the logger for a specific process type
        public void Initialize(string processType, LogLevel logLevel = LogLevel.Info)
        {
            if (_initialized)
                return;
                
            _processType = processType;
            
            // Ensure log directory exists
            Directory.CreateDirectory("/tmp/pxpoint-logs");
            
            // Create log filename with process type and timestamp
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var filename = $"/tmp/pxpoint-logs/pxpoint-{processType}-{timestamp}.log";
            
            try
            {
                _fileWriter = new StreamWriter(filename, append: true, bufferSize: 1024)
                {
                    AutoFlush = true
                };
                
                _initialized = true;
                
                // Log initialization
                LogWithContext(LogLevel.Info, "Logger", "PxPoint logger initialized", 
                    new Dictionary<string, object>
                    {
                        {"process_type", processType},
                        {"log_file", filename}
                    });
            }
            catch (Exception e)
            {
                throw new InvalidOperationException($"Failed to initialize PxPoint logger: {e.Message}", e);
            }
        }
        
        // Log with context and correlation
        public void LogWithContext(LogLevel level, string component, string message,
            Dictionary<string, object> context = null, Dictionary<string, double> performance = null)
        {
            if (!_initialized)
                throw new InvalidOperationException("PxPoint logger not initialized");
            
            var timestamp = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ");
            var correlationId = PxPointCorrelationManager.Instance.GetFullCorrelationId();
            
            // Create structured log entry
            var logEntry = new
            {
                timestamp = timestamp,
                level = level.ToString().ToUpper(),
                process = _processType,
                component = component,
                message = message,
                correlation = correlationId,
                context = context ?? new Dictionary<string, object>(),
                performance = performance ?? new Dictionary<string, double>()
            };
            
            var jsonLog = JsonSerializer.Serialize(logEntry, new JsonSerializerOptions
            {
                WriteIndented = false,
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            });
            
            // Write to file
            _fileWriter.WriteLine(jsonLog);
            
            // Also write to console for immediate feedback
            var consoleMessage = $"[{timestamp}] [{level}] [{component}] {message}";
            if (!string.IsNullOrEmpty(correlationId))
                consoleMessage += $" | correlation:{correlationId}";
            
            Console.WriteLine(consoleMessage);
        }
        
        // Log process milestone events
        public void LogProcessStart(string processType, Dictionary<string, object> config = null)
        {
            var context = config ?? new Dictionary<string, object>();
            context["event_type"] = "process_start";
            context["timestamp"] = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ");
            
            LogWithContext(LogLevel.Info, "Process", $"Process started: {processType}", context);
        }
        
        public void LogProcessEnd(string processType, bool success = true, 
            Dictionary<string, double> metrics = null)
        {
            var context = new Dictionary<string, object>
            {
                ["event_type"] = "process_end",
                ["success"] = success,
                ["timestamp"] = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ")
            };
            
            var logLevel = success ? LogLevel.Info : LogLevel.Error;
            var message = success 
                ? $"Process completed successfully: {processType}"
                : $"Process failed: {processType}";
            
            LogWithContext(logLevel, "Process", message, context, metrics);
        }
        
        // Log activity events
        public void LogActivityStart(string activityName, Dictionary<string, object> context = null)
        {
            var activityContext = context ?? new Dictionary<string, object>();
            activityContext["event_type"] = "activity_start";
            activityContext["timestamp"] = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ");
            
            LogWithContext(LogLevel.Debug, "Activity", $"Activity started: {activityName}", activityContext);
        }
        
        public void LogActivityEnd(string activityName, bool success = true,
            Dictionary<string, double> metrics = null)
        {
            var context = new Dictionary<string, object>
            {
                ["event_type"] = "activity_end",
                ["success"] = success,
                ["timestamp"] = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ")
            };
            
            var logLevel = success ? LogLevel.Debug : LogLevel.Warning;
            var message = success 
                ? $"Activity completed: {activityName}"
                : $"Activity failed: {activityName}";
            
            LogWithContext(logLevel, "Activity", message, context, metrics);
        }
        
        // Log error with full context
        public void LogError(string component, string message, Exception exception = null,
            Dictionary<string, object> context = null)
        {
            var errorContext = context ?? new Dictionary<string, object>();
            errorContext["event_type"] = "error";
            errorContext["timestamp"] = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ");
            
            if (exception != null)
            {
                errorContext["exception"] = exception.ToString();
            }
            
            var errorMessage = $"ERROR: {message}";
            if (exception != null)
            {
                errorMessage += $" | Exception: {exception.Message}";
            }
            
            LogWithContext(LogLevel.Error, component, errorMessage, errorContext);
        }
        
        // Convenience methods
        public void LogInfo(string component, string message, Dictionary<string, object> context = null)
        {
            LogWithContext(LogLevel.Info, component, message, context);
        }
        
        public void LogWarning(string component, string message, Dictionary<string, object> context = null)
        {
            LogWithContext(LogLevel.Warning, component, message, context);
        }
        
        public void LogDebug(string component, string message, Dictionary<string, object> context = null)
        {
            LogWithContext(LogLevel.Debug, component, message, context);
        }
        
        // Shutdown
        public void Shutdown()
        {
            _fileWriter?.Flush();
            _fileWriter?.Close();
            _fileWriter?.Dispose();
        }
    }
    
    public enum LogLevel
    {
        Debug,
        Info,
        Warning,
        Error
    }
}

File: pxpoint/pxpoint_logger.cpp
#include "pxpoint_logger.h"
#include <spdlog/sinks/daily_file_sink.h>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <filesystem>

namespace pxpoint {

PxPointLogger& PxPointLogger::getInstance() {
    static PxPointLogger instance;
    return instance;
}

void PxPointLogger::initialize(const std::string& process_type, spdlog::level::level_enum log_level) {
    if (initialized_) {
        return;
    }
    
    process_type_ = process_type;
    
    // Ensure log directory exists
    std::filesystem::create_directories("/tmp/pxpoint-logs");
    
    // Create log filename with process type and timestamp
    auto now = std::chrono::system_clock::now();
    auto time_t = std::chrono::system_clock::to_time_t(now);
    std::stringstream filename;
    filename << "/tmp/pxpoint-logs/pxpoint-" << process_type << "-" << time_t << ".log";
    
    try {
        // Initialize async thread pool if not already done
        try {
            spdlog::init_thread_pool(8192, 1); // 8192 slots, 1 thread
        } catch (const std::exception&) {
            // Thread pool already initialized
        }
        
        // Create console sink
        auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
        console_sink->set_level(spdlog::level::info);
        console_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%f] [%l] [%n] %v");
        
        // Create daily file sink
        auto file_sink = std::make_shared<spdlog::sinks::daily_file_sink_mt>(
            filename.str(), 0, 0); // No rotation by hour/minute, just daily
        file_sink->set_level(spdlog::level::debug);
        
        // Create JSON pattern for file logging
        file_sink->set_pattern(
            R"({"timestamp":"%Y-%m-%dT%H:%M:%S.%fZ","level":"%l","process":"%n","message":"%v"})");
        
        // Create async logger
        std::vector<spdlog::sink_ptr> sinks{console_sink, file_sink};
        logger_ = std::make_shared<spdlog::async_logger>(
            process_type, 
            sinks.begin(), 
            sinks.end(),
            spdlog::thread_pool(),
            spdlog::async_overflow_policy::block
        );
        
        logger_->set_level(log_level);
        spdlog::register_logger(logger_);
        
        initialized_ = true;
        
        // Log initialization
        logWithContext(spdlog::level::info, "Logger", "PxPoint logger initialized", 
                      {{"process_type", process_type}, {"log_file", filename.str()}});
        
    } catch (const std::exception& e) {
        throw std::runtime_error("Failed to initialize PxPoint logger: " + std::string(e.what()));
    }
}

std::string PxPointLogger::getCurrentTimestamp() {
    auto now = std::chrono::system_clock::now();
    auto time_t = std::chrono::system_clock::to_time_t(now);
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()) % 1000;
    
    std::stringstream ss;
    ss << std::put_time(std::gmtime(&time_t), "%Y-%m-%dT%H:%M:%S");
    ss << '.' << std::setfill('0') << std::setw(3) << ms.count() << 'Z';
    return ss.str();
}

std::string PxPointLogger::formatStructuredMessage(
    const std::string& message,
    const std::unordered_map<std::string, std::string>& context,
    const std::unordered_map<std::string, double>& performance) {
    
    std::stringstream ss;
    ss << message;
    
    // Get correlation information
    auto& correlation_manager = PxPointCorrelationManager::getInstance();
    std::string full_correlation = correlation_manager.getFullCorrelationId();
    
    if (!full_correlation.empty() || !context.empty() || !performance.empty()) {
        ss << " | ";
        
        // Add correlation information
        if (!full_correlation.empty()) {
            ss << "correlation:" << full_correlation;
        }
        
        // Add context fields
        for (const auto& [key, value] : context) {
            ss << " " << key << ":" << value;
        }
        
        // Add performance fields
        for (const auto& [key, value] : performance) {
            ss << " " << key << ":" << std::fixed << std::setprecision(2) << value;
        }
    }
    
    return ss.str();
}

void PxPointLogger::logWithContext(spdlog::level::level_enum level,
                                  const std::string& component,
                                  const std::string& message,
                                  const std::unordered_map<std::string, std::string>& context,
                                  const std::unordered_map<std::string, double>& performance) {
    if (!initialized_) {
        throw std::runtime_error("PxPoint logger not initialized");
    }
    
    std::string structured_message = formatStructuredMessage(message, context, performance);
    
    // Add component to the structured message
    std::stringstream final_message;
    final_message << "[" << component << "] " << structured_message;
    
    logger_->log(level, final_message.str());
}

void PxPointLogger::logProcessStart(const std::string& process_type,
                                   const std::unordered_map<std::string, std::string>& config) {
    auto context = config;
    context["event_type"] = "process_start";
    context["timestamp"] = getCurrentTimestamp();
    
    logWithContext(spdlog::level::info, "Process", 
                  "Process started: " + process_type, context);
}

void PxPointLogger::logProcessEnd(const std::string& process_type, bool success,
                                 const std::unordered_map<std::string, double>& metrics) {
    std::unordered_map<std::string, std::string> context;
    context["event_type"] = "process_end";
    context["success"] = success ? "true" : "false";
    context["timestamp"] = getCurrentTimestamp();
    
    auto log_level = success ? spdlog::level::info : spdlog::level::err;
    std::string message = success ? "Process completed successfully: " : "Process failed: ";
    message += process_type;
    
    logWithContext(log_level, "Process", message, context, metrics);
}

void PxPointLogger::logActivityStart(const std::string& activity_name,
                                    const std::unordered_map<std::string, std::string>& context) {
    auto activity_context = context;
    activity_context["event_type"] = "activity_start";
    activity_context["timestamp"] = getCurrentTimestamp();
    
    logWithContext(spdlog::level::debug, "Activity", 
                  "Activity started: " + activity_name, activity_context);
}

void PxPointLogger::logActivityEnd(const std::string& activity_name, bool success,
                                  const std::unordered_map<std::string, double>& metrics) {
    std::unordered_map<std::string, std::string> context;
    context["event_type"] = "activity_end";
    context["success"] = success ? "true" : "false";
    context["timestamp"] = getCurrentTimestamp();
    
    auto log_level = success ? spdlog::level::debug : spdlog::level::warn;
    std::string message = success ? "Activity completed: " : "Activity failed: ";
    message += activity_name;
    
    logWithContext(log_level, "Activity", message, context, metrics);
}

void PxPointLogger::logError(const std::string& component, const std::string& message,
                            const std::string& exception,
                            const std::unordered_map<std::string, std::string>& context) {
    auto error_context = context;
    error_context["event_type"] = "error";
    error_context["timestamp"] = getCurrentTimestamp();
    if (!exception.empty()) {
        error_context["exception"] = exception;
    }
    
    std::string error_message = "ERROR: " + message;
    if (!exception.empty()) {
        error_message += " | Exception: " + exception;
    }
    
    logWithContext(spdlog::level::err, component, error_message, error_context);
}

void PxPointLogger::shutdown() {
    if (logger_) {
        logger_->flush();
        spdlog::drop(logger_->name());
    }
    
    // Shutdown thread pool
    try {
        spdlog::shutdown();
    } catch (const std::exception&) {
        // Already shutdown or not initialized
    }
}

} // namespace pxpoint